### PR TITLE
first version

### PR DESCRIPTION
**Pull request by dmytrocherepov**

### CODE REVIEW

File: src/main/java/com/example/bookstoreapp/dto/order/OrderDto.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import java.time.LocalDateTime;
+import java.util.List;
+
+public record OrderDto(
+        Long id,
+        Long userId,
+        List<OrderItemDto> orderItems,
+        LocalDateTime orderDate,
+        BigDecimal total,
+        Order.Status status
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/dto/order/OrderDto.java]: U should import `import .../.model.Order.Status;` and then use just Status status field

File: src/main/java/com/example/bookstoreapp/dto/order/OrderRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.example.bookstoreapp.dto.order;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Size;
+
+public record OrderRequestDto(
+        @NotNull
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/dto/order/OrderRequestDto.java]: use @NotBlank instead notNull

File: src/main/java/com/example/bookstoreapp/dto/order/UpdateOrderStatusRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.example.bookstoreapp.dto.order;
+
+import com.example.bookstoreapp.model.Order;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Size;
+
+public record UpdateOrderStatusRequestDto(
+        @NotNull
+        @Size(max = 16)
+        Order.Status status) {
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/dto/order/UpdateOrderStatusRequestDto.java]: The same, u can import the enum from order model

File: src/main/java/com/example/bookstoreapp/mapper/OrderItemMapper.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import com.example.bookstoreapp.config.MapperConfig;
+import com.example.bookstoreapp.dto.order.OrderItemDto;
+import com.example.bookstoreapp.model.CartItem;
+import com.example.bookstoreapp.model.OrderItem;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
+public interface OrderItemMapper {
+    @Mapping(source = "book.price", target = "price")
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/mapper/OrderItemMapper.java:12]: i am not sure, but in my project it does`t work whithout ignoring id
`@Mapping(target = "id", ignore = true)`
! u can ignore this comment

File: src/main/java/com/example/bookstoreapp/mapper/OrderMapper.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import com.example.bookstoreapp.dto.order.OrderDto;
+import com.example.bookstoreapp.model.Order;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class, uses = OrderItemMapper.class)
+public interface OrderMapper {
+    @Mapping(source = "user.id", target = "userId")
+    OrderDto toDto(Order order);
+}
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/mapper/OrderMapper.java:13]: in the future you will have to expand the project, and I think that it would not be bad to add toOrder

Comment (SalnikovMarkiian) [reply src/main/java/com/example/bookstoreapp/mapper/OrderMapper.java:13]: additionaly,  I used OrderMapper for setting totalPrice and orderToItems
Where do u do it ?

File: src/main/java/com/example/bookstoreapp/model/Order.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+@Setter
+@Table(name = "orders")
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true where id = ?")
+@Where(clause = "is_deleted = false")
+@Entity
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/model/Order.java:34]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/example/bookstoreapp/model/Order.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    @JoinColumn(name = "user_id", nullable = false)
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/model/Order.java:35]: @ToString.Exclude
@EqualsAndHashCode.Exclude

add this annotations under User user, if u don`t know what they do - u can google it, or rewatch our lectures

File: src/main/java/com/example/bookstoreapp/model/Order.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    private User user;
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    private Status status;
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/model/Order.java:39]: actually, u can set by default status - NEW, when create order, but up to u

File: src/main/java/com/example/bookstoreapp/model/Order.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
+    @Column(nullable = false)
+    private String shippingAddress;
+    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/model/Order.java:46]: @ToString.Exclude
@EqualsAndHashCode.Exclude

File: src/main/java/com/example/bookstoreapp/model/OrderItem.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+@Where(clause = "is_deleted = false")
+@Entity
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/model/OrderItem.java:31]: @ToString.Exclude
@EqualsAndHashCode.Exclude

File: src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    private final OrderMapper orderMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    @Transactional
+    public OrderDto addOrder(OrderRequestDto orderRequestDto) {
+        User user = getUser();
+        ShoppingCart userShoppingCart = findUserShoppingCart();
+        if (userShoppingCart.getCartItems().isEmpty()) {
+            throw new EntityNotFoundException("Cart is empty");
```
Comment (SalnikovMarkiian) [inline src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java]: why u use  EntityNotFoundException if messege equals "Cart is empty". U have to create custom exception for this case

**Review by SalnikovMarkiian — CHANGES_REQUESTED** [review]
Good solution! 
But u have to fix some comments ^_^

File: src/main/java/com/example/bookstoreapp/controller/BookController.java [hunk 12]
```diff
@@ -27,7 +26,6 @@
 
 @Tag(name = "Book management ", description = "Endpoints for managing books")
 @RestController
-@Validated
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/controller/BookController.java:30]: Why did you delete this annotation from this class?

File: src/main/java/com/example/bookstoreapp/dto/order/OrderRequestDto.java (hunk 2 above)
Comment (liliia-ponomarenko) [reply src/main/java/com/example/bookstoreapp/dto/order/OrderRequestDto.java]: +++

File: src/main/java/com/example/bookstoreapp/repository/shoppingcart/ShoppingCartRepository.java [hunk 13]
```diff
@@ -8,4 +8,5 @@
 @Repository
 public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
     Optional<ShoppingCart> findShoppingCartByUserId(Long id);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/repository/shoppingcart/ShoppingCartRepository.java]:
```suggestion
```

File: src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Pageable;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java:38]:
```suggestion
```

File: src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java (hunk 11 above)
Comment (liliia-ponomarenko) [reply src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java]: agree

File: src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    @Override
+    public OrderItemDto getOrderItemByOrderAndId(Long orderId, Long id) {
+        OrderItem orderItem = itemRepository.findByIdAndOrderId(orderId, id).orElseThrow(
+                () -> new EntityNotFoundException("No such order with id" + id)
+        );
+        return orderItemMapper.toDto(orderItem);
+    }
+
+    private User getUser() {
+        String name = SecurityContextHolder.getContext().getAuthentication().getName();
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java:96]: you can getPrincipal()

File: src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        return userRepository.findByEmail(name).orElseThrow(
+                () -> new EntityNotFoundException("No such user")
+        );
+    }
+
+    private ShoppingCart findUserShoppingCart() {
+        String name = SecurityContextHolder.getContext().getAuthentication().getName();
+        User user = userRepository.findByEmail(name).orElseThrow(
+                () -> new EntityNotFoundException("No such user")
+        );
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java]: Why do you duplicate this logic?

File: src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    private Order createOrder(User user, OrderRequestDto requestDto) {
+        Order order = new Order();
+        order.setOrderDate(LocalDateTime.now());
+        order.setUser(user);
+        order.setShippingAddress(requestDto.shippingAddress());
+        order.setStatus(Order.Status.NEW);
+        order.setTotal(BigDecimal.ZERO);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java:116]: Why do you set total 0?

File: src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+    public List<OrderItemDto> getAllOrderItems(Long id) {
+        return itemRepository.findAllByOrderId(id).stream()
+                .map(orderItemMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public OrderItemDto getOrderItemByOrderAndId(Long orderId, Long id) {
+        OrderItem orderItem = itemRepository.findByIdAndOrderId(orderId, id).orElseThrow(
+                () -> new EntityNotFoundException("No such order with id" + id)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java]:
```suggestion
                () -> new EntityNotFoundException("No such order item with id" + id)
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 19]
```diff
@@ -17,4 +17,9 @@ databaseChangeLog:
       file: db/changelog/changes/08-create-shopping-cart-table.yaml
   - include:
       file: db/changelog/changes/09-create-cartItems-table.yaml
+  - include:
+      file: db/changelog/changes/10-create-order-table.yaml
+  - include:
+      file: db/changelog/changes/11-create-order-items-table.yaml
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
```

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution;)

**Comment by dmytrocherepov** [comment]
Don't know why , but @Validated don't passes on github , while mvn clean package is success

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
this PR seems to contain changes irrelevant to Order model (Dockerfile etc.)

pls either split or find out the reason of irrelevant changes

**Comment by dmytrocherepov** [comment]
> this PR seems to contain changes irrelevant to Order model (Dockerfile etc.)
> 
> pls either split or find out the reason of irrelevant changes

My docker pull-request was approved and I accidentally merged it into this branch

File: src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+    private final OrderMapper orderMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    @Transactional
+    public OrderDto addOrder(OrderRequestDto orderRequestDto) {
+        User user = getUser();
+        ShoppingCart userShoppingCart = findUserShoppingCart();
+        if (userShoppingCart.getCartItems().isEmpty()) {
+            throw new ShoppingCartException("Cart is empty");
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java]:
```suggestion
            throw new OrderCreatingException("Cart is empty");
```

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! But fix one comment before merging;)
